tests/test_pono.py

```diff
@@ -0,0 +1,175 @@
+    itp = ss.create_msat_interpolator()
+
+    # create RTS, populate with encoder
+    rts = pono.RelationalTransitionSystem(solver)
+    pono.CoreIREncoder(top_mod, rts)
+
+    # create BV sorts
+    bvsort16 = solver.make_sort(ss.sortkinds.BV, 16)
+    bvsort2 = solver.make_sort(ss.sortkinds.BV, 2)
+
+    # Reference to inputs
+    a = rts.named_terms["a"]
+    b = rts.named_terms["b"]
+
+    # Make state var equal to output (result of the implementation)
+    imp_res = rts.make_statevar('imp_res', bvsort16)
+    rts.assign_next(imp_res, rts.named_terms["self.c"])
+
+    # Create state var for spec result
+    spec_res = rts.make_statevar('spec_res', bvsort16)
+
+    # Expected op based on config_data param
+    op = [BVAdd, BVSub, BVMul, BVUdiv][config_data]
+
+    # Assign spec result to operation with a and b
+    rts.assign_next(
+        spec_res,
+        solver.make_term(op, a, b)
+    )
+
+    # State machine variable
+    test_state = rts.make_statevar('test_state', bvsort2)
+
+    # Starts at state 0
+    rts.constrain_init(solver.make_term(Equal, test_state,
+                                        solver.make_term(0, bvsort2)))
+
+    # sequential states 0 -> 1 -> 2 -> 3, then stays at 3
+    rts.assign_next(
+        test_state,
+        solver.make_term(
+            Ite,
+            solver.make_term(
+                Equal,
+                test_state,
+                solver.make_term(0, bvsort2)
+            ),
+            solver.make_term(1, bvsort2),
+            solver.make_term(
+                Ite,
+                solver.make_term(
+                    Equal,
+                    test_state,
+                    solver.make_term(1, bvsort2)
+                ),
+                solver.make_term(2, bvsort2),
+                solver.make_term(
+                    Ite,
+                    solver.make_term(
+                        Equal,
+                        test_state,
+                        solver.make_term(2, bvsort2)
+                    ),
+                    solver.make_term(3, bvsort2),
+                    solver.make_term(3, bvsort2)
+                )
+            )
+        )
+    )
+
+    # Boolean sort
+    boolsort = solver.make_sort(ss.sortkinds.BOOL)
+
+    # Create statevar to drive `config_en`
+    config_en = rts.named_terms["config_en"]
+    config_en_driver = rts.make_statevar('config_en_driver',
+                                         boolsort)
+    # Starts low
+    rts.constrain_init(solver.make_term(Equal, config_en_driver,
+                                        solver.make_term(False)))
+    # Goes high for a cycle, then low for rest of states
+    rts.assign_next(
+        config_en_driver,
+        solver.make_term(
+            Ite,
+            solver.make_term(
+                Equal,
+                test_state,
+                solver.make_term(0, bvsort2)
+            ),
+            solver.make_term(True),
+            solver.make_term(
```

Comment:
I think this could just be `solver.make_term(False)` here. Although as mentioned above, this might be more realistic for an automatically generated state machine for driving the reset sequence.

On that note: you'll likely get better performance if you're able to just start the circuit from a concrete reset state. I.e. instead of encoding the reset sequence, just simulate the reset sequence, and assign any of the non-X values in the initial state for formal. That's assuming you're not verifying the reset sequence of course. Not sure how practical that is in general.

Reply:
Hmm, that makes sense, it seems like combining simulation+formal in this way might be a cool technique to explore.  I'll see how we can support it and then we can run some experiments.  I'd imagine for complex systems this might be a good idea, particularly if we can "cache" the reset state based on the design (e.g. we can use the output verilog as a hash, and so long as it doesn't change, we don't need to recompute the reset state).  This would avoid having to redo the logic either in simulation or formal

Reply:
One thing I wasn't sure about, I originally had this code:
```diff
diff --git a/tests/test_pono.py b/tests/test_pono.py
index eb33a79..914eae4 100644
--- a/tests/test_pono.py
+++ b/tests/test_pono.py
@@ -105,7 +105,7 @@ def test_pono(config_data):
                 test_state,
                 solver.make_term(0, bvsort2)
             ),
-            solver.make_term(True),
+            solver.make_term(False),
             solver.make_term(
                 Ite,
                 solver.make_term(
@@ -113,7 +113,7 @@ def test_pono(config_data):
                     test_state,
                     solver.make_term(1, bvsort2)
                 ),
-                solver.make_term(False),
+                solver.make_term(True),
                 solver.make_term(
                     Ite,
                     solver.make_term(
```
Basically, the `config_en` was high a cycle later.  However when I do this, the tests fail.  Is there some relationships between states and cycle position? (e.g. does a state count as a half cycle or full cycle?) Or is there some delay required? (which is why moving up the time of the assertion of config en a state lets the result propagate?)

Reply:
Ahh interesting. My guess is that the issue actually had to do with the property. Since it's not guarded by the value of `test_state` then if you have `config_en` go high a step later, then it can just violated the property in the initial state. You could try changing the `config_en` logic back but then make the property: `test_state >= 1 -> <original prop>`.

Regarding the cycle/state relationship, it depends on the encoding. Right now, the CoreIR encoder detects this automatically by default. So our "abstracted clock" encoding means that every state transition is a full clock cycle. This happens iff there's a single clock and no negedge behavior. Otherwise, each state transition is a "step" and the clock toggling has to be done manually: so normally you'd have a fastest clock that toggles every step and then possibly slower clocks on a different period. The CoreIR encoder also takes a boolean that can force it to abstract the clock (all clocks become synchronized and neg-edge behavior is dropped): https://github.com/upscale-project/pono/blob/157c4c83728b1d6926a5376f0a7a6b04ddf69c4b/frontends/coreir_encoder.h#L15. I haven't tested that feature much though, so let me know if you run into issues.

Reply:
Hmm, I was thinking that the `Distinct` portion of the property took care of that?
```
    prop = pono.Property(
         rts, solver.make_term(Or, solver.make_term(Equal, imp_res, spec_res),
                               solver.make_term(Distinct, test_state,
                                                solver.make_term(3, bvsort2))))
```
unless the initial state is undefined and that portion of the `or` returns false?

Reply:
(I was going for: either state is not 3 or the property must be true

Reply:
Oops, no you're right -- I just didn't look carefully. Hmm could you print the witness and send it to me? After getting `FALSE`, you can just do `print(rts.witness())`

Reply:
Here's the witness for when `config_data = 3` (doing bvudiv)
```
[{(bvadd a (bvneg b)): (_ bv1 16), spec_res.next: (_ bv2 16), (ite (= ((_ extract 0 0) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)) (bvadd a (bvneg b)) (bvadd a b)): (_ bv3 16), config_reg$conf_reg$reg_P_inst0.next: (_ bv0 2), test_state.next: (_ bv1 2), config_reg$conf_reg$reg_P_inst0: (_ bv0 2), config_en_driver.next: false, CLK: false, config_en_driver: false, CLK.next: false, (bvmul a b): (_ bv2 16), (bvudiv a b): (_ bv2 16), (let ((.def_28 (= ((_ extract 0 0) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)))) (ite (= ((_ extract 1 1) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)) (ite .def_28 (bvudiv a b) (bvmul a b)) (ite .def_28 (bvadd a (bvneg b)) (bvadd a b)))): (_ bv3 16), (= ((_ extract 1 1) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)): false, (ite (= ((_ extract 0 0) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)) (bvudiv a b) (bvmul a b)): (_ bv2 16), config_en: false, (ite config_en config_data config_reg$conf_reg$reg_P_inst0): (_ bv0 2), imp_res: (_ bv0 16), b: (_ bv1 16), a: (_ bv2 16), test_state: (_ bv0 2), (bvadd a b): (_ bv3 16), spec_res: (_ bv0 16), (= ((_ extract 0 0) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)): false, config_data: (_ bv3 2), imp_res.next: (_ bv3 16)}, {test_state.next: (_ bv2 2), spec_res.next: (_ bv1 16), imp_res.next: (_ bv9 16), (ite (= ((_ extract 0 0) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)) (bvadd a (bvneg b)) (bvadd a b)): (_ bv9 16), CLK.next: false, CLK: false, config_en_driver.next: true, (bvmul a b): (_ bv20 16), config_reg$conf_reg$reg_P_inst0.next: (_ bv0 2), (bvudiv a b): (_ bv1 16), (bvadd a (bvneg b)): (_ bv1 16), (let ((.def_28 (= ((_ extract 0 0) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)))) (ite (= ((_ extract 1 1) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)) (ite .def_28 (bvudiv a b) (bvmul a b)) (ite .def_28 (bvadd a (bvneg b)) (bvadd a b)))): (_ bv9 16), (= ((_ extract 1 1) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)): false, (ite config_en config_data config_reg$conf_reg$reg_P_inst0): (_ bv0 2), (ite (= ((_ extract 0 0) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)) (bvudiv a b) (bvmul a b)): (_ bv20 16), config_en: false, imp_res: (_ bv3 16), b: (_ bv4 16), a: (_ bv5 16), config_en_driver: false, config_reg$conf_reg$reg_P_inst0: (_ bv0 2), (bvadd a b): (_ bv9 16), test_state: (_ bv1 2), config_data: (_ bv3 2), (= ((_ extract 0 0) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)): false, spec_res: (_ bv2 16)}, {test_state.next: (_ bv3 2), spec_res.next: (_ bv65535 16), imp_res.next: (_ bv0 16), (ite (= ((_ extract 0 0) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)) (bvadd a (bvneg b)) (bvadd a b)): (_ bv0 16), CLK.next: false, CLK: false, config_en_driver.next: false, (bvmul a b): (_ bv0 16), config_reg$conf_reg$reg_P_inst0.next: (_ bv3 2), (bvudiv a b): (_ bv65535 16), (bvadd a (bvneg b)): (_ bv0 16), (let ((.def_28 (= ((_ extract 0 0) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)))) (ite (= ((_ extract 1 1) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)) (ite .def_28 (bvudiv a b) (bvmul a b)) (ite .def_28 (bvadd a (bvneg b)) (bvadd a b)))): (_ bv0 16), (= ((_ extract 1 1) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)): false, (ite config_en config_data config_reg$conf_reg$reg_P_inst0): (_ bv3 2), (ite (= ((_ extract 0 0) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)) (bvudiv a b) (bvmul a b)): (_ bv0 16), config_en: true, imp_res: (_ bv9 16), b: (_ bv0 16), a: (_ bv0 16), config_en_driver: true, config_reg$conf_reg$reg_P_inst0: (_ bv0 2), (bvadd a b): (_ bv0 16), test_state: (_ bv2 2), config_data: (_ bv3 2), (= ((_ extract 0 0) config_reg$conf_reg$reg_P_inst0) (_ bv1 1)): false, spec_res: (_ bv1 16)}]
```

Reply:
I see `config_reg$conf_reg$reg_P_inst0: (_ bv0 2)` and `config_reg$conf_reg$reg_P_inst0: (_ bv0 2)` in the final state of the witness, so it looks like the register value hasn't been updated yet in the final state

Reply:
Hmm, perhaps because `config_en_driver` is a state variable, there's a 1 cycle delay for it to be updated, which in turn takes another cycle for `config_data` to be updated, so we need the extra cycle/state for the register value to be updated

Reply:
Something is strange with `config_en_driver` as well. It looks like it's toggling between values, but it should just be `true` in the first state and then `false` in the rest. I'll try to reproduce this locally.

Reply:
I think you're right. `config_en_driver` doesn't go high until the 3rd state because at state 2, `test_state==1`. Then it takes one more step for the `config_reg` to be updated, so the property can be violated exactly when `test_state` first becomes 3. You could add an extra bit to `test_state` and have it go to 4 before checking the property if you want to test that?

Reply:
Yea, that's essentially what moving the config_en True up a state does.  I'm looking into whether this is an issue for fault, because there are some "combinational" checks done in normal tests, whereas it seems like Pono requires us to only use state variables inside properties. We could just restrict properties at the fault level to be similarly stateful, but is there a way to write a property dependent on something like the next value of a state variable?

Reply:
Right, so properties in Pono are restricted to being over state variables. I know it's somewhat annoying because CoSA was more relaxed about it, but it's much more predictable in terms of behavior. The simplest way to get around that is just to create a property signal, e.g. a state variable `prop` that keeps track of whether the property is satisfied (this extends the trace by 1).

For example, say you want to check if `x.next == x`. Then you just create a variable `xnext_prop` that starts as `True` and then constrain it like: `rts.constrain_trans(solver.make_term(Iff, rts.next(xnext_prop), solver.make_term(Equal, xnext, x)`. And now your property is just `xnext_prop`. I'm planning to handle that automatically in `Property` for you soon (we already handle it automatically in the binary). I can make a PR for that today. I'm not sure I'll stick with the implementation, but the interface should stay the same at least.